File: uiframe/src/main/java/mobi/cangol/mobile/base/BaseNavigationFragmentActivity.java
/*
 * Copyright (c) 2013 Cangol
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mobi.cangol.mobile.base;

import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.view.WindowManager;
import android.widget.FrameLayout;

import java.lang.ref.SoftReference;

import mobi.cangol.mobile.navigation.AbstractNavigationFragmentActivityDelegate;

public abstract class BaseNavigationFragmentActivity extends BaseActionBarActivity {
    public static final String MENU_SHOW = "MENU_SHOW";
    private static final String MENU_TAG = "MenuFragment";
    private SoftReference<BaseMenuFragment> menuFragmentReference;
    private AbstractNavigationFragmentActivityDelegate mHelper;
    protected boolean mFloatActionBarEnabled;

    public boolean isFloatActionBarEnabled() {
        return mFloatActionBarEnabled;
    }

    public void setFloatActionBarEnabled(boolean floatActionBarEnabled) {
        mFloatActionBarEnabled = floatActionBarEnabled;
    }
    /**
     * 返回content布局的id
     *
     * @return
     */
    public abstract int getContentFrameId();

    public void setNavigationFragmentActivityDelegate(AbstractNavigationFragmentActivityDelegate mHelper) {
        this.mHelper = mHelper;
    }

    public AbstractNavigationFragmentActivityDelegate getNavigationFragmentActivityDelegate() {
        return mHelper;
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.getWindow().setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
        mHelper.onCreate(savedInstanceState);
        this.initFragmentStack(getContentFrameId());
    }

    @Override
    public void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        mHelper.attachToActivity(this);
        if (savedInstanceState != null) {
            boolean show = savedInstanceState.getBoolean(MENU_SHOW);
            mHelper.showMenu(show);
        }
    }

    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(MENU_SHOW, isShowMenu());
    }
    @Override
    public void setBackgroundColor(int color) {
        super.setBackgroundColor(color);
        mHelper.setBackgroundColor(color);
    }
    @Override
    public void setBackgroundResource(int resId) {
        super.setBackgroundResource(resId);
        mHelper.setBackgroundResource(resId);
    }

    @Override
    public View findViewById(int id) {
        View v = super.findViewById(id);
        if (v != null)
            return v;
        return mHelper.getRootView().findViewById(id);
    }

    @Override
    public void setContentView(int id) {
        setContentView(getLayoutInflater().inflate(id, null));
    }

    @Override
    public void setContentView(View v) {
        setContentView(v, new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));
    }

    @Override
    public FrameLayout getMaskView() {
        if(isFloatActionBarEnabled()){
           return mHelper.getMaskView();
        }else{
           return super.getMaskView();
        }
    }
    @Override
    public void displayMaskView(boolean show) {
        if(isFloatActionBarEnabled()){
            mHelper.displayMaskView(show);
        }else{
            super.displayMaskView(show);
        }
    }

    public void showMenu(boolean show) {
        mHelper.showMenu(show);
    }

    public boolean isShowMenu() {
        return mHelper.isShowMenu();
    }

    public void setMenuEnable(boolean enable) {
        mHelper.setMenuEnable(enable);
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        boolean b = mHelper.onKeyUp(keyCode, event);
        if (b) return b;
        return super.onKeyUp(keyCode, event);
    }

    public int getCurrentModuleId() {
        if (menuFragmentReference == null||menuFragmentReference.get()==null) {
            throw new IllegalStateException("menuFragment is null");
        } else {
            return menuFragmentReference.get().getCurrentModuleId();
        }
    }

    public BaseMenuFragment getMenuFragment() {
        if(menuFragmentReference!=null)
            return menuFragmentReference.get();
        else
            return null;
    }

    public void setCurrentModuleId(int moduleId) {
        if (menuFragmentReference != null&&menuFragmentReference.get()!=null) {
            menuFragmentReference.get().setCurrentModuleId(moduleId);
        }
    }

    public final void setMenuFragment(Class<? extends BaseMenuFragment> fragmentClass, Bundle args) {
        BaseMenuFragment menuFragment = (BaseMenuFragment) Fragment.instantiate(this, fragmentClass.getName(), args);
        menuFragmentReference=new SoftReference<>(menuFragment);
        FragmentTransaction t = this.getSupportFragmentManager()
                .beginTransaction();
        t.replace(mHelper.getMenuFrameId(), menuFragment, MENU_TAG);
        t.commitAllowingStateLoss();
        getSupportFragmentManager().executePendingTransactions();
    }

    public final void setContentFragment(Class<? extends BaseFragment> fragmentClass, String tag, Bundle args, int moduleId) {
        replaceFragment(fragmentClass, tag, args);
        setCurrentModuleId(moduleId);
    }

    public final void setContentFragment(Class<? extends BaseFragment> fragmentClass, String tag, Bundle args) {
        replaceFragment(fragmentClass, tag, args);
    }
    public final void setContentFragment(Class<? extends BaseFragment> fragmentClass, Bundle args) {
        replaceFragment(fragmentClass, fragmentClass.getName(), args);
    }
    public void notifyMenuOnClose() {
        if (menuFragmentReference != null&&menuFragmentReference.get()!=null)
            menuFragmentReference.get().onClosed();
    }

    public void notifyMenuOnOpen() {
        if (menuFragmentReference != null&&menuFragmentReference.get()!=null)
            menuFragmentReference.get().onOpen();
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        BaseMenuFragment menuFragment = (BaseMenuFragment) getSupportFragmentManager().findFragmentByTag(MENU_TAG);
        if(menuFragment!=null){
            menuFragmentReference=new SoftReference<>(menuFragment);
        }
    }
}

File: uiframe/src/main/kotlin/mobi/cangol/mobile/base/BaseNavigationFragmentActivity.kt
package mobi.cangol.mobile.base

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.KeyEvent
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.FrameLayout
import mobi.cangol.mobile.navigation.AbstractNavigationFragmentActivityDelegate
import java.lang.ref.WeakReference


abstract class BaseNavigationFragmentActivity : BaseActionBarActivity() {
    private var menuFragmentReference: WeakReference<BaseMenuFragment>? = null
    private var mHelper: AbstractNavigationFragmentActivityDelegate? = null
    private var mFloatActionBarEnabled: Boolean = false

    open fun isFloatActionBarEnabled(): Boolean {
        return mFloatActionBarEnabled
    }

    open fun setFloatActionBarEnabled(floatActionBarEnabled: Boolean) {
        mFloatActionBarEnabled = floatActionBarEnabled
    }

    /**
     * 返回content布局的id
     *
     * @return
     */
    abstract fun getContentFrameId(): Int

    fun setNavigationFragmentActivityDelegate(mHelper: AbstractNavigationFragmentActivityDelegate) {
        this.mHelper = mHelper
    }

    fun getNavigationFragmentActivityDelegate(): AbstractNavigationFragmentActivityDelegate {
        return mHelper!!
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.window.setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
        mHelper?.onCreate(savedInstanceState)
        this.initFragmentStack(getContentFrameId())
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        mHelper?.attachToActivity(this)
        if (savedInstanceState != null) {
            val show = savedInstanceState.getBoolean(MENU_SHOW)
            mHelper?.showMenu(show)
        }
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(MENU_SHOW, isShowMenu())
    }

    override fun setBackgroundColor(color: Int) {
        super.setBackgroundColor(color)
        mHelper?.setBackgroundColor(color)
    }

    override fun setBackgroundResource(resId: Int) {
        super.setBackgroundResource(resId)
        mHelper?.setBackgroundResource(resId)
    }

    override fun <T : View> findViewById(id: Int): T? {
        val v = super.findViewById<T>(id)
        return v ?: mHelper!!.getRootView().findViewById(id)
    }

    override fun setContentView(id: Int) {
        setContentView(layoutInflater.inflate(id, null))
    }

    override fun setContentView(v: View) {
        setContentView(v, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
    }

    override fun getMaskView(): FrameLayout {
        return if (isFloatActionBarEnabled()) {
            mHelper!!.getMaskView()
        } else {
            super.getMaskView()
        }
    }

    override fun displayMaskView(show: Boolean) {
        if (isFloatActionBarEnabled()) {
            mHelper!!.displayMaskView(show)
        } else {
            super.displayMaskView(show)
        }
    }

    fun showMenu(show: Boolean) {
        mHelper!!.showMenu(show)
    }

    fun isShowMenu(): Boolean {
        return mHelper!!.isShowMenu()
    }

    fun setMenuEnable(enable: Boolean) {
        mHelper!!.setMenuEnable(enable)
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        val b = mHelper!!.onKeyUp(keyCode, event)
        return if (b) b else super.onKeyUp(keyCode, event)
    }

    fun getCurrentMoudleId(): Int {
        return if (menuFragmentReference == null || menuFragmentReference!!.get() == null) {
            throw IllegalStateException("menuFragment is null")
        } else {
            menuFragmentReference!!.get()!!.getCurrentModuleId()
        }
    }

    fun setCurrentModuleId(moduleId: Int) {
        check(!(menuFragmentReference == null || menuFragmentReference!!.get() == null)) { "menuFragment is null" }
        menuFragmentReference!!.get()?.setCurrentModuleId(moduleId)
    }

    fun getMenuFragment(): BaseMenuFragment? {
        return menuFragmentReference!!.get()
    }

    fun setMenuFragment(fragmentClass: Class<out BaseMenuFragment>, args: Bundle?) {
        val menuFragment = Fragment.instantiate(this, fragmentClass.name, args) as BaseMenuFragment
        menuFragmentReference = WeakReference(menuFragment)
        val t = this.supportFragmentManager
                .beginTransaction()
        t.replace(mHelper!!.getMenuFrameId(), menuFragment, MENU_TAG)
        t.commitAllowingStateLoss()
        supportFragmentManager.executePendingTransactions()
    }

    fun setContentFragment(fragmentClass: Class<out BaseFragment>, tag: String, args: Bundle?, moduleId: Int) {
        replaceFragment(fragmentClass, tag, args)
        setCurrentModuleId(moduleId)
    }

    fun setContentFragment(fragmentClass: Class<out BaseFragment>, tag: String, args: Bundle?) {
        replaceFragment(fragmentClass, tag, args)
    }

    fun setContentFragment(fragmentClass: Class<out BaseFragment>, args: Bundle?) {
        replaceFragment(fragmentClass, fragmentClass.name, args)
    }

    fun notifyMenuOnClose() {
            menuFragmentReference?.get()?.onClosed()
    }

    fun notifyMenuOnOpen() {
        if (menuFragmentReference != null && menuFragmentReference!!.get() != null)
            menuFragmentReference?.get()?.onOpen()
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        val menuFragment = supportFragmentManager.findFragmentByTag(MENU_TAG) as BaseMenuFragment?
        if (menuFragment != null) {
            menuFragmentReference = WeakReference(menuFragment)
        }
    }

    companion object {
        private const  val MENU_SHOW = "MENU_SHOW"
        private const val MENU_TAG = "MenuFragment"
    }
}